arrows/core/parameterize_mesh.cxx

```diff
@@ -0,0 +1,294 @@
+        AC = -pt2pt3;
+        longest_edge = 2;
+      }
+      // transform the face to 2d
+      vector_2d a(0.0, 0.0);
+      vector_2d b(AB.norm(), 0.0);
+      double proj = AC.dot(AB) / AB.norm();
+      vector_2d c(proj, (AC - proj * AB.normalized()).norm());
+      // scale B and C by the resolution
+      b /= d_->resolution;
+      c /= d_->resolution;
+
+      total_area += 0.5 * (b(0) * c(1));
+      unsigned int w = static_cast<unsigned int>(std::floor(std::max(b(0), c(0)))) + 1;
+      unsigned int h = static_cast<unsigned int>(std::floor(c(1))) + 1;
+      if (longest_edge == 0)
+      {
+        // pt1 is A, pt2 is B, pt3 is C
+        triangles[f] = {a, b, c, f, h, w};
+      }
+      else if (longest_edge == 1)
+      {
+        // pt1 is C, pt2 is A, pt3 is B
+        triangles[f] = {c, a, b, f, h, w};
+      }
+      else if (longest_edge == 2)
+      {
+        // pt1 is B, pt2 is C, pt3 is A
+        triangles[f] = {b, c, a, f, h, w};
+      }
+    }
+  }
+
+  // Sort triangles by height
+  std::vector<unsigned int> face_indices(triangles.size(), 0);
+  std::iota(face_indices.begin(), face_indices.end(), 0);
+
+  std::sort(face_indices.begin(), face_indices.end(), [&triangles](unsigned int i, unsigned int j)
+  {
+    return triangles[i].height < triangles[j].height;
+  });
+
+  // Pack triangles
+  std::vector<vector_2d> tcoords(mesh->num_faces() * 3);
+
+  // approximate max_width to have a more or less square texture atlas
+  const unsigned int max_width = static_cast<unsigned int>(std::ceil(sqrt(total_area * 2)));
+
+  unsigned int current_u = d_->exterior_margin;
+  unsigned int current_v = d_->exterior_margin;
+  unsigned int next_v = current_v;
+  vital::vector_2d shift(0.0, 0.0);
+  unsigned int max_u = 0, max_v = 0;
+  for (unsigned int f : face_indices)
+  {
+    if (current_u + triangles[f].width + d_->exterior_margin >= max_width)
+    {
+      current_u = d_->exterior_margin;
+      current_v = next_v + d_->interior_margin;
+    }
+    shift(0) = current_u;
+    shift(1) = current_v;
+    triangles[f].a += shift;
+    triangles[f].b += shift;
+    triangles[f].c += shift;
+
+    if (current_u + triangles[f].width > max_u)
+      max_u = current_u + triangles[f].width;
+    if (current_v + triangles[f].height > max_v)
+      max_v = current_v + triangles[f].height;
+
+    next_v = std::max(next_v, current_v + triangles[f].height);
+    current_u += triangles[f].width + d_->interior_margin;
+  }
+
+  unsigned int width = max_u + d_->exterior_margin;
+  unsigned int height = max_v + d_->exterior_margin;
+
+  // Normalize texture coordinates
+  vector_2d norm(width, height);
+  for (unsigned int i = 0; i < triangles.size(); ++i)
+  {
+    tcoords[i * 3 + 0] = triangles[i].a.cwiseQuotient(norm);
+    tcoords[i * 3 + 1] = triangles[i].b.cwiseQuotient(norm);
+    tcoords[i * 3 + 2] = triangles[i].c.cwiseQuotient(norm);
+  }
+  mesh->set_tex_coords(tcoords);
+
+  return std::make_pair(width, height);
```

Comment:
I don't think you need to return this.  The texture coordinates should be normalized to a square space between 0 and 1 in each axis.  If later you want to figure out what size texture you need (e.g. 512x512, 1024x1024, etc.) to obtain a certain resolution, you can easily pick one or more triangles and compare the ratio of the square root of their areas in both world and texture space to figure out the size needed to achieve a certain resolution.